src/build_mesh_tree.py

```diff
@@ -0,0 +1,190 @@
+import argparse
+import os
+import pickle
+import sys
+import time
+from typing import List
+
+import pandas as pd
+import tqdm
+from pubmedpy.eutilities import esearch_query
+
+ARTICLE_THRESHOLD = 200000
+CACHE_FILE = 'heading.tmp'
+
+class Node():
+    def __init__(self, heading: str, id: str):
+        self.heading = heading
+        self.id = id
+        self.children = []
+
+    def add_child(self, child: "Node"):
+        self.children.append(child)
+
+
+def persist_to_file(file_name):
+    # https://stackoverflow.com/questions/16463582/memoize-to-disk-python-persistent-memoization
+    def decorator(original_func):
+        try:
+            cache = pickle.load(open(file_name, 'rb'))
+        except (IOError, ValueError):
+            cache = {}
+
+        def new_func(param):
+            if param not in cache:
+                cache[param] = original_func(param)
+                pickle.dump(cache, open(file_name, 'wb'))
+            return cache[param]
+
+        return new_func
+    return decorator
+
+
+def make_path_safe(path: str) -> str:
+    # https://stackoverflow.com/questions/7406102/create-sane-safe-filename-from-any-unsafe-string
+    return "".join([c for c in path if c.isalpha() or c.isdigit() or c==' ']).rstrip()
+
+
+@persist_to_file(CACHE_FILE)
+def get_heading_article_count(heading: str) -> pd.DataFrame:
+    """
+    Return the number of unique dois that fall under the given MeSH heading
+
+    Arguments
+    ---------
+    heading: The MeSH heading to examine
+
+    Returns
+    -------
+    count: The number or dois for the heading
+
+    Note:
+    This function is based on code from Le et al., and used in accordance with their license:
+    https://github.com/greenelab/iscb-diversity/blob/master/02.process-pubmed.ipynb
+    """
+    payload = {
+        'db': 'pubmed',
+        'term': f'"journal article"[pt] AND "{heading}"[MeSH Terms] AND English[Language]',
+    }
+
+    retry_interval = 1
+    while retry_interval < 5000:
+        try:
+            pubmed_ids = esearch_query(payload)
+            return len(pubmed_ids)
+        except Exception as e:
+            print(f'Retrying after {retry_interval} seconds due to {e}', file=sys.stderr)
+            time.sleep(retry_interval)
+            retry_interval *= 4
+
+    raise RuntimeError(f'Repeated timeouts when downloading {heading}', file=sys.stderr)
+
+
+
+def get_headings(current_node: Node) -> List[str]:
+    children = current_node.children
+    print(current_node.heading)
```

Comment:
Do you want to print this, or was this for debugging?

Reply:
Yeah, I left that one in to track progress/get an idea of what MeSH headings exist